src/foam/lib/parse/Grammar.js

```diff
@@ -0,0 +1,150 @@
+/**
+ * @license
+ * Copyright 2020 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+ foam.CLASS({
+  package: 'foam.lib.parse',
+  name: 'Grammar',
+
+  javaImports: [
+    'foam.lib.parse.Parser',
+    'foam.lib.parse.StringPStream',
+    'java.util.Map',
+    'foam.lib.parse.SymbolParser'
+  ],
+
+  requires: [
+    'foam.parse.ParserWithAction'
+  ],
+
+  properties: [
+    {
+      class: 'Map',
+      name: 'symbols',
+      javaFactory: 'return new java.util.HashMap<String, foam.lib.parse.Parser>();'
+    }
+  ],
+
+  methods: [
+    {
+      name: 'addSymbol',
+      args: [
+        { name: 'symName', type: 'String' },
+        { name: 'symParser', type: 'Parser' }
+      ],
+      javaCode: `
+      getSymbols().put(symName, symParser);
+      `
+    },
+    {
+      name: 'parseString',
+      type: 'foam.lib.parse.PStream',
+      args: [
+        { name: 'str', type: 'String' },
+        { name: 'optName', type: 'String' },
+        { name: 'x',  javaType: 'foam.lib.parse.ParserContext' }
```

Comment:
Should I need to supply the parserContext, or should it automatically create one for me?